DRAFT_USAGE.md

```diff
@@ -151,15 +185,70 @@ proc handler(request: Request) {.async.} =
   await request.resp("Oops!, 404", code = Http404)
 ```
 
-## Authentication HttpCode401
+## Basic Authentication
+
+Incomplete API. This is subject to change. See src/scorper/http/httpbasicauth
+
+A request can have authorization basic headers automatically decoded and then passed to a validator for verification.
+
 ```nim
+# Create validator
+# Must take a Request and two string params and return a Future[bool]
+
+proc basicValidation(request: Request, user, pass: string): Future[bool] {.async.} =
+  # Check the decoded user and pass against your database/whatever
+  # return true for authentication to be accepted and for the calling callback to continue
+  # return false for automatic Http401 response
+  return true
+
+# Create your callback
 proc handler(request: Request) {.async.} =
-  if req.headers.hasKey("Authorization"):
-    await req.resp("")              # Client will receive a Http200
+  # Create validation object and pass to request method basicAuth
+  var validator = HttpBasicAuthValidator(basicValidation) # our validator proc
+  if await request.basicAuth(validator):
+    # Do things because request is validified
   else:
-    await req.respBasicAuth()       # Client will receive a Http401
+    # Do things here because request was invalid
+    # RESPONSE UNNECESSARY; see note below
+
+  # basicAuth proc returns a bool; however any failure to validate the auth header
+  # will result in an automatic response of either Http400 or Http401. An if/else
+  # statement can be used however responses for failure are unnecessary.
+  # Using the statement below is therefore perfectly valid
+  #
+  #   if not await request.basicAuth(validator): return
+
+# Can create router here or set server handler directly
+let address, flags = "127.0.0.1:8888", {ReuseAddr}
+let server = newScorper(address, handler, flags)
+server.start()
+waitFor server.join()
+```

```

Comment:
Is this use of validators correct?

Reply:
sorry, I don't understand, where  you called "validators" ?

Reply:
```nim
# Create validator
# Must take a Request and two string params and return a Future[bool]

proc basicValidation(request: Request, user, pass: string): Future[bool] {.async.} =
  # Check the decoded user and pass against your database/whatever
  # return true for authentication to be accepted and for the calling callback to continue
  # return false for automatic Http401 response
  return true

# Create your callback
proc handler(request: Request) {.async.} =
  # Create validation object and pass to request method basicAuth
  var validator = HttpBasicAuthValidator(basicValidation) # our validator proc
  if await request.basicAuth(validator):
    # Do things because request is validified
  else:
    # Do things here because request was invalid
    # RESPONSE UNNECESSARY; see note below

  # basicAuth proc returns a bool; however any failure to validate the auth header
  # will result in an automatic response of either Http400 or Http401. An if/else
  # statement can be used however responses for failure are unnecessary.
  # Using the statement below is therefore perfectly valid
  #
  #   if not await request.basicAuth(validator): return

# Can create router here or set server handler directly
let address, flags = "127.0.0.1:8888", {ReuseAddr}
let server = newScorper(address, handler, flags)
server.start()
waitFor server.join()
```